precompiles/collective/src/lib.rs

```diff
@@ -0,0 +1,281 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Precompile to interact with pallet_collective instances.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use core::marker::PhantomData;
+use fp_evm::{Precompile, PrecompileOutput};
+use frame_support::{
+	dispatch::Dispatchable,
+	sp_runtime::traits::Hash,
+	traits::ConstU32,
+	weights::{GetDispatchInfo, PostDispatchInfo},
+};
+use pallet_evm::AddressMapping;
+use precompile_utils::prelude::*;
+use sp_core::{Decode, H256};
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+/// Solidity selector of the Executed log.
+pub const SELECTOR_LOG_EXECUTED: [u8; 32] = keccak256!("Executed(bytes32)");
+
+/// Solidity selector of the Proposed log.
+pub const SELECTOR_LOG_PROPOSED: [u8; 32] = keccak256!("Proposed(address,uint32,bytes32,uint32)");
+
+/// Solidity selector of the Voted log.
+pub const SELECTOR_LOG_VOTED: [u8; 32] = keccak256!("Voted(address,bytes32,bool)");
+
+type GetProposalLimit = ConstU32<{ 2u32.pow(16) }>;
+
+#[generate_function_selector]
+#[derive(Debug, PartialEq)]
+pub enum Action {
+	Execute = "execute(bytes)",
+	Propose = "propose(uint32,bytes)",
+	Vote = "vote(bytes32,uint32,bool)",
+	Close = "close(bytes32,uint32,uint64)",
+	ProposalHash = "proposalHash(bytes)",
```

Comment:
Is there a reason why we dont provide other storage accesors? (Voting, Members, etc)

Reply:
I only mirrored calls that can be made by non-roots. We could have accessors, however getting "Voting" is not possible yet due to private fields.

Reply:
The ones I see more valuable are "members", "ProposalOf" and "Proposals", but we can add them later